### ISSUE TITLE
[Autograd] Missing higher-order gradients block molecular force training

### ISSUE THREAD

**Issue by leeleolay** [PaddlePaddle/Paddle#79493]
# [Autograd] Missing higher-order gradients block molecular force training

## Environment

- PaddlePaddle: 3.1.0
- Wheel commit: `cdcd76a93d1dfe767dde768198e1d7c7488b93f6`
- Latest `develop` inspected: `f92f172716c6190c36573057cf16750d8b1c07b9`
- Python: 3.10
- Device: CPU
- Platform: Linux x86_64

## Use case

An atomistic model predicts energy `E(R, theta)`, computes forces as
`F = -dE/dR`, and trains against force labels. The force loss differentiates
the coordinate gradient with respect to model parameters, so every operator on
the energy path must support higher-order automatic differentiation.

`paddle.atan2`, `paddle.linalg.cross`, `paddle.nn.Embedding`, and
`paddle.scatter_nd_add` currently prevent this workflow.

## Minimal reproduction

Save the following as `reproduce.py`. Run each case in a separate process
because the `cross` case terminates the interpreter.

```python
import argparse

import paddle


def atan2_case():
    x = paddle.to_tensor([0.4, -0.7], dtype="float32")
    y = paddle.to_tensor([1.2, 0.8], dtype="float32")
    x.stop_gradient = False
    y.stop_gradient = False
    loss = paddle.atan2(x, y).square().sum()
    return paddle.grad(loss, [x, y], create_graph=True)


def cross_case():
    x = paddle.to_tensor(
        [[0.2, 0.4, -0.5], [0.7, -0.3, 0.6]], dtype="float32"
    )
    y = paddle.to_tensor(
        [[-0.1, 0.8, 0.3], [0.5, 0.2, -0.4]], dtype="float32"
    )
    x.stop_gradient = False
    loss = paddle.linalg.cross(x, y).square().sum()
    return paddle.grad(loss, x, create_graph=True)


def embedding_case():
    paddle.seed(42)
    embedding = paddle.nn.Embedding(8, 4)
    linear = paddle.nn.Linear(8, 4)
    atom_ids = paddle.to_tensor([1, 3], dtype="int64")
    coordinates = paddle.to_tensor(
        [[0.2, 0.4, -0.5, 0.1], [0.7, -0.3, 0.6, 0.2]],
        dtype="float32",
    )
    coordinates.stop_gradient = False
    features = paddle.concat([embedding(atom_ids), coordinates], axis=-1)
    energy = linear(features).square().sum()
    return paddle.grad(energy, coordinates, create_graph=True)


def scatter_nd_add_case():
    updates = paddle.to_tensor(
        [[0.2, 0.4], [-0.5, 0.1], [0.7, -0.3]], dtype="float32"
    )
    updates.stop_gradient = False
    index = paddle.to_tensor([[0], [1], [0]], dtype="int64")
    aggregated = paddle.scatter_nd_add(
        paddle.zeros([2, 2], dtype="float32"), index, updates
    )
    loss = aggregated.square().sum()
    return paddle.grad(loss, updates, create_graph=True)


CASES = {
    "atan2": atan2_case,
    "cross": cross_case,
    "embedding": embedding_case,
    "scatter_nd_add": scatter_nd_add_case,
}

parser = argparse.ArgumentParser()
parser.add_argument("case", choices=CASES)
args = parser.parse_args()
paddle.set_device("cpu")
print(f"paddle={paddle.__version__}, device=cpu", flush=True)
print(CASES[args.case](), flush=True)
```

```bash
python reproduce.py atan2
python reproduce.py cross
python reproduce.py embedding
python reproduce.py scatter_nd_add
```

## Actual behavior

| API | Result with `create_graph=True` |
| --- | --- |
| `paddle.atan2` | `atan2_grad` has no grad op |
| `paddle.linalg.cross` | CPU SIGSEGV in `CrossGradKernel` |
| `paddle.nn.Embedding` | `embedding_grad` has no grad op |
| `paddle.scatter_nd_add` | `scatter_nd_add_grad` has no grad op |

The embedding reproducer deliberately joins embedding and coordinate branches
before a nonlinear energy. A disconnected or purely linear lookup does not
exercise the force-training graph.

The current `develop` operator YAML still has first-gradient declarations for
these APIs but no corresponding double-gradient declarations. This statement is
based on source inspection, not an unbuilt development wheel.

## Expected behavior

`paddle.grad(..., create_graph=True)` should construct a differentiable first
gradient so a subsequent force loss can backpropagate to model parameters. An
unsupported higher-order derivative should report an exception and must not
segfault.

## Current workarounds

- Replace `atan2` with differentiable cosine/sine components and a detached
  forward-value correction.
- Expand the cross product into elementwise multiplication and subtraction.
- Replace embedding lookup with one-hot matrix multiplication for force models.
- Replace scatter reduction with one-hot matrix multiplication for force models.

These workarounds match the reference PyTorch forward values and parameter
gradients, but increase memory use and require framework-specific branches.

A focused patch adding `scatter_nd_add_double_grad` and a CPU dygraph unit test
was submitted as [PaddlePaddle/Paddle#79494](https://github.com/PaddlePaddle/Paddle/pull/79494).
